### ISSUE TITLE
`port-writes-special?` ⊂"running under Dr Racket"

### ISSUE THREAD

**Issue by greghendershott** [racket/htdp#234]
The assumption here seems to be that, is if the output port can write "specials", then we're running under Dr Racket:

https://github.com/racket/htdp/blob/aef9ae1c8e4c1e256edde5ab842d20d558202ee2/htdp-lib/htdp/bsl/runtime.rkt#L112-L136

However that's not necessarily true. Racket Mode presents an output port that accepts `write-special`. Therefore it ends up getting these `number-markup` structs. It had been surprised by these, ergo https://github.com/greghendershott/racket-mode/issues/732.

I merged an accommodation for that, today. But it seems like `port-writes-special?` isn't a sufficient test, here, for "port that knows how to handle `number-markup` structs, and will end up in a snip in the DrRacket GUI"?  I don't know enough to propose a better test, but I wanted to ask.

**Comment by rfindler** [comment]
I agree that the `port-writes-special?` predicate isn't precise enough; I'm not sure what's the best thing to do in general, however.

For the specific case of the teaching languages and the markup that gets written, there is an API that will print these things in an ascii mode (drawing boxes and whatnot). It would be nice if we could arrange that this got used.

But I don't think it makes sense for that to be something that happens in Racket mode; it would be nice if Racket mode's port could just be like "no thanks" and there would be a fallback to the ascii. I'm really stuck here mentally as I can't get past "there should be a change to Racket's ports" which really isn't a very good direction, just as a general rule!

**Comment by greghendershott** [comment]
It seems this is about the "host" environment supplying the port, rather than the port per se.

A mechanism that occurs to me is [`read-language` `get-info`](https://docs.racket-lang.org/tools/lang-languages-customization.html). Although typically that's used to get lang info to the host, there's also a preexisting example of the REPL submit predicate key, which is a function in the lang called by the host.

In that vein: Could that `use-number-markup?` function in `bsl/runtime.rkt` instead be a parameter (or function with parameter-like signature) exposed via a new `get-info` key? It could default to false (good for things like command-line `racket`, Racket Mode, etc.). DrRacket could set it true.

(I'm not sure what to name the key, because I don't understand how structs in specials like `number-markup` end up as objects in GUI "snips". A straw man name: `'simple-tree-lib-markup-structs`.)

---


> there is an API that will print these things in an ascii mode (drawing boxes and whatnot). It would be nice if we could arrange that this got used.

IIUC So then maybe the parameter isn't a boolean. Maybe there are three values: "plain", "ascii fallback", "full structs". Whatever you think makes sense. In any case, still the basic idea of using `get-info` to let the host environment pick which (instead of the lang trying to guess which).
